dandi/files/__init__.py

```diff
@@ -0,0 +1,288 @@
+    "dandi_file",
+    "find_dandi_files",
+    "find_bids_dataset_description",
+]
+
+lgr = get_logger()
+
+
+def find_dandi_files(
+    *paths: str | Path,
+    dandiset_path: Optional[str | Path] = None,
+    allow_all: bool = False,
+    include_metadata: bool = False,
+) -> Iterator[DandiFile]:
+    """
+    Yield all DANDI files at or under the paths in ``paths`` (which may be
+    either files or directories).  Files & directories whose names start with a
+    period are ignored.  Directories are only included in the return value if
+    they are of a type represented by a `LocalDirectoryAsset` subclass, in
+    which case they are not recursed into.
+
+    :param dandiset_path:
+        The path to the root of the Dandiset in which the paths are located.
+        All paths in ``paths`` must be equal to or subpaths of
+        ``dandiset_path``.  If `None`, then the Dandiset path for each asset
+        found is implicitly set to the parent directory.
+    :param allow_all:
+        If true, unrecognized assets and the Dandiset's :file:`dandiset.yaml`
+        file are returned as `GenericAsset` and `DandisetMetadataFile`
+        instances, respectively.  If false, they are not returned at all.
+    :param include_metadata:
+        If true, the Dandiset's :file:`dandiset.yaml` file is returned as a
+        `DandisetMetadataFile` instance.  If false, it is not returned at all
+        (unless ``allow_all`` is true).
+    """
+
+    # A pair of each file or directory being considered plus the most recent
+    # BIDS dataset_description.json file at the path (if a directory) or in a
+    # parent path
+    path_queue: deque[tuple[Path, Optional[BIDSDatasetDescriptionAsset]]] = deque()
+    for p in map(Path, paths):
+        if dandiset_path is not None:
+            try:
+                p.relative_to(dandiset_path)
+            except ValueError:
+                raise ValueError(
+                    "Path {str(p)!r} is not inside Dandiset path {str(dandiset_path)!r}"
+                )
+        path_queue.append((p, None))
+    while path_queue:
+        p, bidsdd = path_queue.popleft()
+        if p.name.startswith("."):
+            continue
+        if p.is_dir():
+            if p.is_symlink():
+                lgr.warning("%s: Ignoring unsupported symbolic link to directory", p)
+            elif dandiset_path is not None and p == Path(dandiset_path):
+                if (p / BIDS_DATASET_DESCRIPTION).exists():
+                    bids2 = dandi_file(p / BIDS_DATASET_DESCRIPTION, dandiset_path)
+                    assert isinstance(bids2, BIDSDatasetDescriptionAsset)
+                    bidsdd = bids2
+                path_queue.extend((q, bidsdd) for q in p.iterdir())
+            elif any(p.iterdir()):
+                try:
+                    df = dandi_file(p, dandiset_path, bids_dataset_description=bidsdd)
+                except UnknownAssetError:
+                    if (p / BIDS_DATASET_DESCRIPTION).exists():
```

Comment:
> may be just that condition above could be adjusted that e.g. we first condition on having BIDS_DATASET_DESCRIPTION and then something happens "differently" depending on either it is a top of dandiset or not?

That would lead to wrong behavior if a Zarr contained a `dataset_description.json`.

> Also may be it would allow to catch/report an unsupported ATM (AFAIK?) case when BIDS dataset is embedded somewhere within DANDI dataset but there is no BIDS_DATASET_DESCRIPTION on top.

If there's no `dataset_description.json`, how would we know there's a BIDS dataset there?

Reply:
Comment added: a654c8b.

Reply:
> > Also may be it would allow to catch/report an unsupported ATM (AFAIK?) case when BIDS dataset is embedded somewhere within DANDI dataset but there is no BIDS_DATASET_DESCRIPTION on top.
> 
> If there's no `dataset_description.json`, how would we know there's a BIDS dataset there?

I was not clear.  In " there is no BIDS_DATASET_DESCRIPTION on top" I meant that it is "on top of the dandiset". I.e. when we have `dandiset.yaml` and `rawdata/dataset_description.json` but no `dataset_description.json`, i.e. BIDS dataset is embedded without entire dandiset being a BIDS dataset.